pennylane/_qubit_device.py

```diff
@@ -0,0 +1,207 @@
+                self.apply(operation)
+
+            self.post_apply()
+
+            # Pass queue of observables to pre_measure
+            self.pre_measure()
+
+            for obs in observables:
+                # Pass instances directly
+                if obs.return_type is Expectation:
+                    results.append(self.expval(obs))
+
+                elif obs.return_type is Variance:
+                    results.append(self.var(obs))
+
+                elif obs.return_type is Sample:
+                    results.append(np.array(self.sample(obs)))
+
+                elif obs.return_type is Probability:
+                    results.append(list(self.probability(wires=obs.wires).values()))
+
+                elif obs.return_type is not None:
+                    raise QuantumFunctionError("Unsupported return type specified for observable {}".format(obs.name))
+
+            self.post_measure()
+
+            self._op_queue = None
+            self._obs_queue = None
+            self._parameters = None
+
+            # Ensures that a combination with sample does not put
+            # expvals and vars in superfluous arrays
+            if all(obs.return_type is Sample for obs in observables):
+                return self._asarray(results)
+            if any(obs.return_type is Sample for obs in observables):
+                return self._asarray(results, dtype="object")
+
+            return self._asarray(results)
+
+    def rotate_basis(self, observable):
+        """Rotates the specified wires such that they
+        are in the eigenbasis of the provided observable.
+
+        Args:
+            observable (Observable): the observable that is to be
+                measured
+        """
+        for diag_gate in observable.diagonalizing_gates()[::-1]:
+            self.apply(diag_gate)
+
+    def expval(self, observable):
+        wires = observable.wires
+        par = observable.parameters
+
+        if self.analytic:
+            # exact expectation value
+            eigvals = observable.eigvals
+            prob = np.abs(self._state ** 2)
+            prob = self.marginal_prob(prob, wires=wires)
+            return (eigvals @ prob).real
+
+        # estimate the ev
+        return np.mean(self.sample(observable))
+
+    def var(self, observable):
+        wires = observable.wires
+        par = observable.parameters
+
+        if self.analytic:
+            # exact variance value
+            eigvals = observable.eigvals
+            prob = np.abs(self._state ** 2)
+            prob = self.marginal_prob(prob, wires=wires)
+            return (eigvals ** 2) @ prob - (eigvals @ prob).real ** 2
+
+        return np.var(self.sample(observable))
+
+    def sample(self, observable):
+        if isinstance(observable, qml.Identity):
+            return np.ones([self.shots])
+        self.rotate_basis(observable)
+        # Sampling is supported on software simulators
+        # HW devices will have to provide their own sample() method
```

Comment:
It might be worth overwriting the ``Device.sample`` docstring here, so that this is explicit

Reply:
As per our discussion, this will be TODO for the sampling PR